[config] Support new 'site' option in Infra Agent

olivielpeau opened the pull request:
### What does this PR do?

Adds support of new 'site' option in Infra Agent (for now, other agent components will follow).

* Overriden by `dd_url` if set
* Should be fully backward compatible
* `dd_url` is not necessarily set on `config.Datadog` anymore. The "resolved" URL for the infra Agent should now be accessed with `config.GetMainInfraEndpoint`. @DataDog/agent-integrations: this means any integration that's trying to access the `dd_url` with `config.get("dd_url")` will not necessarily get a value back, is that an issue?

### Motivation

Allows setting the DD site to which the Agent should send data with one single config option that can apply to all agent components (not implemented yet on other components).

### Additional Notes

**Look at the first commit for the actual changes needed for the feature.**

The other commits are not absolutely necessary, but fix various small things (see commit descriptions).

The last commit (https://github.com/DataDog/datadog-agent/pull/2412/commits/3fc2beed1d5c9241108191567564a42379e6d874) is an attempt to make the package easier to unit-test, borne out of frustration of not being able to cleanly test this change :) I can definitely extract it out in a separate PR though.

TODO in separate PRs:
- [ ] support in logs
- [x] support in trace: https://github.com/DataDog/datadog-trace-agent/pull/485
- [x] support in process: https://github.com/DataDog/datadog-process-agent/pull/192
- [x] support for `DD_SITE` in the install scripts (should be merged after 6.6.0 is out): https://github.com/DataDog/datadog-agent/pull/2414

olivielpeau committed:
[config] Add new 'site' option for Infra Agent

Overriden by `dd_url` if latter is set. Should be fully backward
compatible.

olivielpeau committed:
[flare] Fix domain used to send flares

We should use a versioned domain. This was broken since 6.5.0 and
ec90952dc8f0c6bc68f7c809618c1b0770c1fb65

It didn't prevent flares from being sent though.

olivielpeau committed:
[config] Remove unused test fixtures

olivielpeau committed:
[legacy] Make config importer ignore dd_url if left to default

To encourage users not to use 'dd_url' anymore, and to prefer using
'site' if needed.

olivielpeau committed:
[Dockerfiles] Remove explicit setting of DD_DD_URL in entrypoints

`dd_url` is an optional setting in Agent 6, no need to always set it
(and it's actually counter-productive now that we offer the `site`
option)

olivielpeau committed:
release note

olivielpeau committed:
[windows] Allow passing `SITE` option on the command line

xvello commented on pkg/config/config.go:
```diff
@@ -73,7 +78,9 @@ func init() {
-	BindEnvAndSetDefault("dd_url", "https://app.datadoghq.com")
+	// Don't set a default on 'site' to allow detecting with viper whether it's set in config
+	Datadog.BindEnv("site")
+	Datadog.BindEnv("dd_url")
```
IIRC viper leverages the default value to configure the expected type as we set [SetTypeByDefaultValue](https://godoc.org/github.com/spf13/viper#SetTypeByDefaultValue) to true. What could be the side-effects of not setting the type?

Would a default of `""` like some other options work?

olivielpeau replied:
thanks for mentioning this, wasn't aware of this (and the unit tests aren't using our viper wrapper at the moment). Let me fix the unit tests and give this a try 👍

olivielpeau replied:
tried this: unfortunately `Datadog.IsSet` will always return `true` as long as a default was set on the config key. So AFAIK, as long as a default value is set on a config option, viper doesn't offer any way to check whether a config option was set by the user, which is a problem in this feature because we want to log a message if both `site` and `dd_url` are set by the user.

Not setting a default does mean that `Datadog.Get("site")` doesn't necessarily return a string though.

Alternative: we continue setting a default of the empty string on `site` and `dd_user`, and don't make a difference anymore between the case where the option isn't set by the user, and the case where the option is set by the user to an empty-string-like value.

xvello replied:
default type is only used in the generic `Get()` method that we do not use https://github.com/spf13/viper/blob/62edee319679b6ceaec16de03b966102d2dea709/viper.go#L663
I think it should be safe to work without, as long as we explictly call `GetString` as we're used to do

olivielpeau committed:
[config] Use new Config interface in site/dd_url/endpoints unit tests

olivielpeau committed:
[config] Improve testability of package

* avoid mutating global Datadog config from arbitrary functions
* always pass a Config to all these functions

gbbr commented on pkg/config/types.go:
```diff
@@ -59,4 +59,8 @@ type Config interface {
 
 	BindPFlag(key string, flag *pflag.Flag) error
+
+	// Datadog specific
```
I'd say better write actual documentation instead of this. "Datadog specific" is not helping understand this better. Are the rest not Datadog specific? Or, create an embedded interface which is called something else, and document that as being Datadog specific, saying what each method does, if you think this section will grow.

olivielpeau replied:
you're right, the comment doesn't mean anything as it is, I'll fix this

gbbr reviewed (commented):
Just a small question/suggestion.

sfluor commented on pkg/config/config.go:
```diff
@@ -486,14 +490,29 @@ func AddAgentVersionToDomain(domain string, app string) (string, error) {
 }
 
+// getMainEndpoint implements the logic to extract the main DD URL from a config
+func getMainEndpoint(config Config) (ddURL string) {
```
Could it be possible to export it publicly and add an `agent prefix` as a parameter so it can be used from the other agents ?

olivielpeau replied:
that'd make sense yes 👍  I'll have a look

olivielpeau replied:
This ended up being more complex than what you described: see `GetMainEndpoint` in https://github.com/DataDog/datadog-agent/pull/2412/commits/9fb323939da96f5de67d32b044a83cfc9d83335a 

Even with this change, the `DD_PROCESS_AGENT_URL` env var would still need to be handled specifically in process_agent because it doesn't follow the regular env var naming scheme.

Let me know if you think this commit is useful (I'll remove it if it's not), or if you have any other comments!

sfluor replied:
Thanks ! Since the logic for handling env variables and file variables is separated in the process-agent I think it would still be useful to use this helper for the file config 👍

olivielpeau replied:
@sfluor sounds good. Please note that `GetMainEndpoint` takes into account the value of the `DD_SITE` env var (and the `DD_PROCESS_CONFIG_PROCESS_DD_URL` as well, even if this one's not really relevant)

olivielpeau committed:
Attempt at making the logic usable by other components than infra

olivielpeau committed:
Clarify separation between API implemented by Viper and our helpers

Also, make `BindEnvAndSetDefault` exported again, since it definitely
makes sense to use it outside the package as well.

gbbr commented on pkg/config/types.go:
```diff
@@ -59,8 +65,13 @@ type Config interface {
+// DDConfigHelpers describes the config helpers that are not provided by viper.Viper
+type DDConfigHelpers interface {
+	// GetConfigEnvVars returns a list of the non-sensitive env vars that the config supports
 	GetConfigEnvVars() []string
```
I'd suggest (as a good practice) to always consider the name of the package to be part of exported identifiers. That way, it will be easier to come up with good names. Since this package is already called `config` it might not make sense to include the word "config" in identifiers. For example:
* prefer `EnvVars` or `GetEnvVars` (to match Viper) to `GetConfigEnvVars`
* prefer `Env` to `DDConfigHelpers` (helpers is a very generic catch-all term which will be attractive every time someone doesn't know where to put somewhere). This interface mostly addresses environment variables so it will be ok to call it `Env`
* prefer `Viper` to `ViperConfig`

At the same time (contradicting my previous comment), this could all be part of the same interface. No reason to have two now, right? The fact that that interface would also implement Viper's is just a side effect.

olivielpeau replied:
that's a good point overall, thanks for commenting on this. I'll make some naming changes. To me, separating the 2 interfaces is just a way to explicitly point out what's implemented by viper and what's not, as we're in the process of investigating if/how we could replace viper, so I'll leave most of it as it is.

olivielpeau replied:
getting confused by the back and forth 😅 : I'll keep only one interface, rename things, and clarify the overall intent in comments

gbbr replied:
Sorry for creating confusion. My intention is purely to have a nice API that is easy to understand to potential newcomers. Thanks for being patient.

olivielpeau replied:
no apologies needed, your feedback was very welcome :)

gbbr commented on pkg/config/viper.go:
```diff
@@ -248,20 +258,15 @@ func (c *safeConfig) BindPFlag(key string, flag *pflag.Flag) error {
-// and tracks existing config env vars
-func (c *safeConfig) bindEnvAndSetDefault(key string, val interface{}) {
+// BindEnvAndSetDefault implements the DDConfigHelpers interface
+func (c *safeConfig) BindEnvAndSetDefault(key string, val interface{}) {
```
This seems like purely a convenience method. Do we really need it?

olivielpeau replied:
it's very useful (and we use it for most config options) so I'd say we need it yes

gbbr commented on pkg/config/types.go:
```diff
@@ -59,8 +65,13 @@ type Config interface {
-	bindEnvAndSetDefault(key string, val interface{})
+	// BindEnvAndSetDefault sets the default value for a config parameter and adds an env binding
+	// in one call
+	BindEnvAndSetDefault(key string, val interface{})
```
It looks like this just calls `BindEnv` and `SetDefault` internally, so it might be that we could just remove it and instead just add this other method (GetEnvVars) into the original interface. That'd remove a lot of extra complications imho.

gbbr reviewed (commented):
I just have a couple of design suggestions. They are purely optional and I don't feel strongly about them, but I do think they would simplify things a bit.

olivielpeau committed:
Merge branch 'master' into olivielpeau/new-site-option

Fix conflict in pkg/config/viper.go

achntrl commented on pkg/config/viper.go:
```diff
@@ -16,243 +16,267 @@ import (
+		key := input[0]
+		if !strings.Contains(key, "_key") {
+			envVarName := strings.Join([]string{c.envPrefix, strings.ToUpper(key)}, "_")
+			c.configEnvVars = append(c.configEnvVars, envVarName)
```
Shouldn't the `c.RLock()` be changed to a `c.Lock()` ?

olivielpeau replied:
the lock was originally only meant to make calls to viper thread-safe, but you're right in this case we should fully `Lock`

hush-hush commented on pkg/config/viper.go:
```diff
@@ -16,243 +16,267 @@ import (
 func (c *safeConfig) BindEnv(input ...string) error {
 	c.RLock()
 	defer c.RUnlock()
+	if len(input) == 1 {
```
Why do it only when input length is 1 ?

olivielpeau replied:
this logic that tracks which env vars are configurable pre-existed, I just moved it around. It only works when `BindEnv` is used with one argument (in this case `BindEnv` deduces the env var key from the name of the config option, for example it adds a `DD_` prefix, and our logic later makes the same assumptions)

hush-hush replied:
Could point me to it ? I can't find where we pulled this from (should we remove it from where it is ?) ?

I also think that since we now have a wrapper around viper we should update this logic with to handle every arguments. I'm afraid we will forget about it and it will come back as a bug later.

olivielpeau replied:
this logic was added here: https://github.com/DataDog/datadog-agent/pull/2294 (originally in `config.go`, I removed it from there)

I'll move the related tests to `viper_test.go` and add a FIXME

olivielpeau replied:
done, resolving this

hush-hush commented on pkg/config/viper.go:
```diff
@@ -16,243 +16,267 @@ import (
 	defer c.RUnlock()
+	if len(input) == 1 {
+		key := input[0]
+		if !strings.Contains(key, "_key") {
```
Could we add a comment to explain this line. I'm guessing it's to avoid shipping the app_key/api_key in the flare ?

I think we should move this logic out of the config package into the flare.

olivielpeau replied:
I agree. This logic pre-existed though, happy to address this in a separate PR

hush-hush replied:
Ok, let's create a card in the backlog then. The flare package might already remove `*_key` when dumping env variables.

hush-hush commented on pkg/config/config.go:
```diff
@@ -70,257 +74,255 @@ type Proxy struct {
-	BindEnvAndSetDefault("auth_token_file_path", "")
-	BindEnvAndSetDefault("bind_host", "localhost")
+	// Don't set a default on 'site' to allow detecting with viper whether it's set in config
+	config.BindEnv("site")
```
Should we have a default for site ? Like we had for dd_url ?
What if people don't have `dd_url` in their configuration and upgrade to a new version of the agent ?

olivielpeau replied:
We can't set the default of `site` here because of the comment above (`to allow detecting with viper whether it's set in config`) . Same reason for `dd_url` actually (I'll update the comment).

I _think_ the PR description (that I updated slightly) answers your last question:

>`dd_url` is not necessarily set on `config.Datadog` anymore. The "resolved" URL for the infra Agent should now be accessed with `config.GetMainInfraEndpoint`.

So if users don't have `dd_url` set in their config, `config.Datadog.GetString("dd_url")` will now return an empty value, but the Agent will still use all the expected defaults because it now accesses the default endpoint with `config.Datadog.GetMainInfraEndpoint`.

hush-hush replied:
Why move the default logic from the viper to the code ?

We could add a default to `site` and remove the elseif/else here https://github.com/DataDog/datadog-agent/pull/2412/files#diff-091ace02b49a46f55f2586cfec89a160R513

What do you think ?

olivielpeau replied:
if we set a default with Viper directly, we can't know anymore whether the user actually set the option or not in their config/env vars (see convo there: https://github.com/DataDog/datadog-agent/pull/2412#discussion_r222917900). Since we want to display a log message when `site` and `dd_url` are both set, we need to know whether the user actually set these settings

hush-hush replied:
I see, ok :+1:

hush-hush commented on pkg/config/config.go:
```diff
@@ -446,14 +448,24 @@ func Load() error {
+	return getMainInfraEndpoint(Datadog)
+}
+
+// GetMainEndpoint returns the main DD URL defined in the config, based on `site` and the prefix, or ddURLKey
```
Where is this used ? I don't see it.

olivielpeau replied:
it's used by process agent (see https://github.com/DataDog/datadog-process-agent/pull/192), could also be used by logs for example. Do you think it's worth making it more clear in the comment?

hush-hush replied:
:+1:

hush-hush commented on pkg/config/config.go:
```diff
@@ -486,14 +498,33 @@ func AddAgentVersionToDomain(domain string, app string) (string, error) {
+}
+
+// getMainEndpoint implements the logic to extract the DD URL from a config, based on `site` and ddURLKey
+func getMainEndpoint(config Config, prefix string, ddURLKey string) (resolvedDDURL string) {
```
Will we ever use this function with another `ddUrlKey` than `dd_url` ? Same thing for `prefix` and `defaultInfraURLPrefix`.

I feel like we could remove `getMainInfraEndpoint` function and directly call `getMainEndpoint` with just the `config` parameter.

olivielpeau replied:
(see my previous comment :) )

hush-hush commented on pkg/config/config.go:
```diff
@@ -26,11 +26,15 @@ import (
+// DefaultSite is the default site the Agent sends data to.
+const DefaultSite = "datadoghq.com"
+
+const defaultInfraURLPrefix = "https://app."
```
Do we want to be able to change the prefix using the configuration ?

hush-hush replied:
I'm ok with using `dd_url`, let's rename the variable then.

olivielpeau committed:
Address review comments

olivielpeau committed:
Merge branch 'master' into olivielpeau/new-site-option

ajacquemot reviewed (commented):
Coming up after the battle but everything looks good to me though
